docs/enhancements/auto-return-agent-to-infra-env.md

```diff
@@ -0,0 +1,187 @@
+---
+title: auto-return-agent-to-infra-env
+authors:
+  - "@carbonin"
+creation-date: 2022-05-17
+last-updated: 2022-06-02
+---
+
+# Automatically Return Agents to InfraEnv
+
+## Summary
+
+Without a BareMetalHost there is currently no way to automate the process
+of unbinding a host from a cluster. When a user wants to reuse a host for a new
+cluster they have to use the assisted-service unbind API, then manually boot
+the host back into the discovery image.
+
+To support automating this procedure, the assisted-service will configure nodes
+in installed clusters to run an agent. When appropriate this agent will
+reconfigure the host to boot into a discovery image, allowing the host to
+be made available for installation without the user manually interacting with
+a management system to reboot the host into the discovery image.
+
+## Motivation
+
+A user creating and managing clusters using the assisted-service may not have
+access to the systems used to configure and boot hosts. This is especially
+true in bare-metal use cases when a BareMetalHost is not present. The
+assisted-service can make this use case easier by handling the reprovisioning
+process.
+
+### Goals
+
+* Reconfigure a host to boot the discovery image as a part of unbinding from a
+  cluster without a BareMetalHost
+
+### Non-Goals
+
+* Automate graceful removal of a node from an existing cluster
+* Booting a node that is unresponsive or not currently part of an installed
+  cluster
+* Registering an agent to an InfraEnv other than the one it is currently
+  associated with
+
+## Proposal
+
+When a host should be returned to the InfraEnv use the kubeconfig created when
+the cluster was installed to run an agent on the host as a systemd service 
+using a MachineConfig. This agent will handle downloading the required 
+artifacts and rebooting the host into the discovery process.
+
+This process will only be used in situations where the assisted-service is sure
+to have connectivity to the installed cluster's API endpoint. In cases such as
+the SaaS this feature will not be used.
+
+### Timeline
+
+1. User unbinds the host from the cluster
+2. Assisted service creates a MachineConfig to run the 
```

Comment:
Would a DaemonSet have permissions to do what we need (remount /boot as read/write, add a boot entry, and reboot)?
Also do DaemonSets get removed if someone drains a node?

Reply:
I'm genuinely curious about this.
I did a look around the documentation, but don't see anything indicating that any special permissions are granted to DaemonSet pods.

The only thing I did find was some information about how they're scheduled slightly differently, but nothing that would answer my question about draining. (Although I'll probably just test this in a bit after I get a cluster installed).

Reply:
The `oc adm drain` command has a `--ignore-daemonsets` flag that's important to use, because otherwise the drain won't remove daemonset pods.

It just runs a Pod; can we give a Pod the needed permissions? I would think so unless there's a gotcha.

Reply:
Also after looking more into machineconfigs it looks like the pool concept will make this difficult to target something to run only on a single node.

Specifically these bits of documentation:
> MCO applies changes to operating systems in pools of machines.
> A node can have multiple labels applied that indicate its type, such as master or worker, however it can be a member of only a single machine config pool

Maybe it won't be too hard to get the agent to run as a very privileged daemonset pod :pray: